### ISSUE TITLE
[Tutorial] Infer your own data

### ISSUE THREAD

**Issue by LyWangPX** [YuanGongND/gopt#11]
# Infer your own data

This tutorial is aimed at helping people understand how to infer any data with GOPT.
There will be some restrictions on "any data", I will explain it later.

1. Install Kaldi and GOPT. Sometimes Kaldi can be fuzzy. If meeting any problems, use a docker image of Kaldi instead.
2. Download the original `speechocean762` to your disk.
3. Copy `speechocean762` and change to your dataset’s name. In this example, I use `test_dataset`
4. There are multiple hacks we need to do to make our `test_dataset` runnable.
5. First, delete unnecessary .wav files. Replace them with your own.
6. Update all files in both test and train. Including `spk2age`, `spk2gender`, `spk2utt`, `text`, `utt2spk`, `wav.scp`. Remember, words in text need to be capitalized.
7. Because I only have 1 wav file, all these files will be 1-line. For example, I set the speaker ID to be `0001` and utt_id to be `test`. So:
    
    `wav.scp` includes
    
    ```
    test	WAVE/SPEAKER0001/test.wav
    ```
    
    `spk2utt` includes
    
    ```
    0001 test
    ```
    
    You can infer how to write the rest.
    
8. In `resource/text-phone`, delete unnecessary lines and replace your own. Here, each lines begins with \<utt_id\>.\<n\> which represents the n-th word in your text. After it, please append the corresponding phones of that word. 
    
    To be specific, find your corresponding phones in `resource/lexicon.txt`. For instance, the word FAN would be `F AE0 N`. For all the first phones, add an additional suffix B, for the last phones, add an additional suffix E and for all others, add the suffix I.
    
    For example, if my text is “FAN WORKS”, the final result in text-phone is
    
    ```
    test.0 F_B AE0_I N_E
    test.1 W_B ER0_I K_I S_E
    ```
    
    If your don’t do this right, you will stuck at stage 8.
    
9. Download and extract all tars in [https://kaldi-asr.org/models/m13](https://kaldi-asr.org/models/m13)
10. In `gop_speechocean762/s5/run.sh` , change lines 38-42 to your extracted results. In my case, I use
    
    ```
    librispeech_eg=../../librispeech/s5
    model=$librispeech_eg/exp/chain_cleaned/tdnn_1d_sp
    ivector_extractor=$librispeech_eg/exp/nnet3_cleaned/extractor
    lang=$librispeech_eg/data/lang_test_tgsmall
    ```
    
    Also, change `stage` to `2`to avoid download. Change `nj` to the number of examples. In my case, I set to 1.
    
11. After running run.sh, confirm you have non-empty feat files in `gop_train` and `gop_test`. Mine looks like this 
    
    ```
    (gopt) yifan@XXX:~/develop/kaldi/egs/gop_speechocean762/s5/exp/gop_train$ tree .
    .
    ├── feat.1.ark
    ├── feat.1.scp
    ├── feat.scp
    ├── gop.1.ark
    ├── gop.1.scp
    ├── gop.scp
    └── log
        └── compute_gop.1.log
    ```
    
12. Execute the following original guide in GOPT.
    
    ```
    kaldi_path=your_kaldi_path
    cd $gopt_path
    mkdir -p data/raw_kaldi_gop/librispeech
    cp src/extract_kaldi_gop/{extract_gop_feats.py,extract_gop_feats_word.py} ${kaldi_path}/egs/gop_speechocean762/s5/local/
    cd ${kaldi_path}/egs/gop_speechocean762/s5
    ```
    
13. Now we need to change the original GOPT files
14. First, in `extract_gop_feats.py`, delete the continue [https://github.com/YuanGongND/gopt/blob/master/src/extract_kaldi_gop/extract_gop_feats.py#L54](https://github.com/YuanGongND/gopt/blob/master/src/extract_kaldi_gop/extract_gop_feats.py#L54). (PS: `label` in this file is `lable`. Ummm, you can’t unsee them.)
15. In the same file, because we do not have score, change [https://github.com/YuanGongND/gopt/blob/master/src/extract_kaldi_gop/extract_gop_feats.py#L61](https://github.com/YuanGongND/gopt/blob/master/src/extract_kaldi_gop/extract_gop_feats.py#L61) to
    
    ```
    lables.append([ph])
    ```
    
16. Run the edited `python local/extract_gop_feats.py`, skip `extract_gop_feats_word.py`, not needed for inference.
17. Continue with 
    
    ```
    cd $gopt_path
    cp -r ${kaldi_path}/egs/gop_speechocean762/s5/gopt_feats/* data/raw_kaldi_gop/<your dataset name>
    ```
    
18. Change another GOPT file, `src/prep_data/gen_seq_data_phn.py`. Because we do not have score any more, all we want to have is the phn. Also we need to replace the hardcoding path to \<your dataset name\>. You can debug it yourself, here is my edited results.
    
    ```
    # -*- coding: utf-8 -*-
    # @Time    : 9/19/21 11:13 PM
    # @Author  : Yuan Gong
    # @Affiliation  : Massachusetts Institute of Technology
    # @Email   : yuangong@mit.edu
    # @File    : gen_seq_data_phn.py
    
    # Generate sequence phone input and label for seq2seq models from raw Kaldi GOP features.
    
    import numpy as np
    
    def load_feat(path):
        file = np.loadtxt(path, delimiter=',')
        return file
    
    def load_keys(path):
        file = np.loadtxt(path, delimiter=',', dtype=str)
        return file
    
    def load_label(path):
        file = np.loadtxt(path, delimiter=',', dtype=str)
        return file
    
    def process_label(label):
        pure_label = []
        for i in range(0, label.shape[0]):
            pure_label.append(float(label[i, 1]))
        return np.array(pure_label)
    
    def process_feat_seq(feat, keys, labels, phn_dict):
        key_set = []
        for i in range(keys.shape[0]):
            cur_key = keys[i].split('.')[0]
            key_set.append(cur_key)
    
        feat_dim = feat.shape[1] - 1
    
        utt_cnt = len(list(set(key_set)))
        print('In total utterance number : ' + str(utt_cnt))
    
        # Pad all sequence to 50 because the longest sequence of the so762 dataset is shorter than 50.
        seq_feat = np.zeros([utt_cnt, 50, feat_dim])
        # -1 means n/a, padded token
        # [utt, seq_len, 0] is the phone label, and the [utt, seq_len, 1] is the score label
        seq_label = np.zeros([utt_cnt, 50, 2]) - 1
    
        # the key format is utt_id.phn_id
        prev_utt_id = keys[0].split('.')[0]
    
        row = 0
        for i in range(feat.shape[0]):
            cur_utt_id, cur_tok_id = keys[i].split('.')[0], int(keys[i].split('.')[1])
            # if a new sequence, start a new row of the feature vector.
            if cur_utt_id != prev_utt_id:
                row += 1
                prev_utt_id = cur_utt_id
    
            # The first element is the phone label.
            seq_feat[row, cur_tok_id, :] = feat[i, 1:]
    
            # [utt, seq_len, 0] is the phone label
            print(labels)
            seq_label[row, cur_tok_id, 0] = phn_dict[labels[i]]
            # [utt, seq_len, 1] is the score label, range from 0-2
            # seq_label[row, cur_tok_id, 1] = labels[i, 1]
    
        return seq_feat, seq_label
    
    def gen_phn_dict(label):
        phn_dict = {}
        phn_idx = 0
        for i in range(label.shape[0]):
            if label[i] not in phn_dict:
                phn_dict[label[i]] = phn_idx
                phn_idx += 1
        return phn_dict
    
    # generate sequence training data
    tr_feat = load_feat('../../data/raw_kaldi_gop/test_dataset/tr_feats.csv')
    tr_keys = load_keys('../../data/raw_kaldi_gop/test_dataset/tr_keys_phn.csv')
    tr_label = load_label('../../data/raw_kaldi_gop/test_dataset/tr_labels_phn.csv')
    phn_dict = gen_phn_dict(tr_label)
    print(phn_dict)
    tr_feat, tr_label = process_feat_seq(tr_feat, tr_keys, tr_label, phn_dict)
    print(tr_feat.shape)
    print(tr_label.shape)
    np.save('../../data/seq_data_test_dataset/tr_feat.npy', tr_feat)
    np.save('../../data/seq_data_test_dataset/tr_label_phn.npy', tr_label)
    
    # generate sequence test data
    te_feat = load_feat('../../data/raw_kaldi_gop/test_dataset/te_feats.csv')
    te_keys = load_keys('../../data/raw_kaldi_gop/test_dataset/te_keys_phn.csv')
    te_label = load_label('../../data/raw_kaldi_gop/test_dataset/te_labels_phn.csv')
    te_feat, te_label = process_feat_seq(te_feat, te_keys, te_label, phn_dict)
    print(te_feat.shape)
    print(te_label.shape)
    np.save('../../data/seq_data_test_dataset/te_feat.npy', te_feat)
    np.save('../../data/seq_data_test_dataset/te_label_phn.npy', te_label)
    ```
    
19. The last step requires you to run these lines. Skip word and utterence.

    ```
    mkdir data/seq_data_<your dataset name>
    cd src/prep_data
    python gen_seq_data_phn.py
    ```

20. Finally, in `gopt/data/<your dataset name>`, you will see kindly two files that are needed to do the inference. `te_feat.npy` and `te_label_phn.npy`. But remember, the `te_label_phn.npy`contains both `phn` and `scores` which we have not generated (and not needed). So, in order to do the inference, run the following.
    
    PS: to simplify stuff, my train dataset is the same as test dataset.
    
    ```python
    
    import torch
    import sys
    import os
    sys.path.append(os.path.abspath('../src/'))
    from models import GOPT
    gopt = GOPT(embed_dim=24, num_heads=1, depth=3, input_dim=84)
    # GOPT is trained with dataparallel, so it need to be wrapped with dataparallel even you have a single gpu or cpu
    gopt = torch.nn.DataParallel(gopt)
    sd = torch.load('gopt_librispeech/best_audio_model.pth', map_location='cpu')
    gopt.load_state_dict(sd, strict=True)
    
    import numpy as np
    input_feat = np.load("<your_path>/te_feat.npy")
    input_phn = np.load("<your_path>/te_label_phn.npy")
    t_input_feat = torch.from_numpy(input_feat)
    t_phn = torch.from_numpy(input_phn[:,:,0])
    gopt = gopt.float()
    gopt.eval()
    with torch.no_grad():
        print(gopt(t_input_feat.float(),t_phn.float()))
    ```
    

Good Luck!
Restrictions: If your text contains words that are not in lexicon, you are out of luck.

**Comment by YuanGongND** [comment]
Hi there, 

Thanks so much for sharing this! I think this would be super helpful. 

I want to ask that if you are interested in pull this to the main branch (e.g., as a `md.` file under the main directory)?

-Yuan

**Comment by LyWangPX** [comment]
Of course! Let me fix some typos and make the PR.

令和5年1月27日(金) 1:59 Yuan Gong ***@***.***>:

**Comment by YuanGongND** [comment]
Thanks so much. I just merged your pull to the main branch. 

This is the question asked most commonly by the users, so your contribution would definitely be appreciated.

-Yuan

**Comment by LyWangPX** [comment]
Thanks!
Done.

**Comment by amandeepbaberwal** [comment]
Hi @LyWangPX i followed your guide, on step 20 after i ran the script i get this result. Could you please explain what these scores mean?

`(tensor([[1.7884]]), tensor([[1.4905]]), tensor([[1.7742]]), tensor([[1.7411]]), tensor([[1.7569]]), tensor([[[1.3277],
         [1.1209],
         [1.2091],
         [1.2541],
         [1.2698],
         [1.1702],
         [1.1688],
         [1.1946],
         [1.1906],
         [1.2431],
         [0.9490],
         [1.0390],
         [1.0029],
         [1.0249],
         [1.1234],
         [1.0179],
         [0.9685],
         [1.0225],
         [0.9370],
         [1.1267],
         [1.0197],
         [0.9029],
         [0.9954],
         [1.0375],
         [1.0546],
         [0.9973],
         [1.0303],
         [1.0476],
         [0.9626],
         [1.0615],
         [0.8320],
         [0.9930],
         [0.9372],
         [0.8351],
         [0.9818],
         [1.0020],
         [0.9363],
         [0.9362],
         [1.0583],
         [1.0371],
         [0.9932],
         [1.0459],
         [1.0267],
         [1.0154],
         [1.0279],
         [1.0406],
         [0.9935],
         [0.9378],
         [1.0589],
         [0.9972]]]), tensor([[[0.3930],
         [0.0982],
         [0.1626],
         [0.1966],
         [0.2523],
         [0.1397],
         [0.1701],
         [0.1465],
         [0.2313],
         [0.2835],
         [1.1706],
         [1.2555],
         [1.2182],
         [1.2374],
         [1.3283],
         [1.2313],
         [1.2000],
         [1.2312],
         [1.1628],
         [1.3359],
         [1.2384],
         [1.1439],
         [1.1986],
         [1.2622],
         [1.2589],
         [1.1860],
         [1.2566],
         [1.2814],
         [1.2020],
         [1.2586],
         [1.0791],
         [1.2016],
         [1.1631],
         [1.0782],
         [1.1928],
         [1.2277],
         [1.1717],
         [1.1726],
         [1.2802],
         [1.2533],
         [1.2286],
         [1.2571],
         [1.2550],
         [1.2379],
         [1.2475],
         [1.2501],
         [1.2199],
         [1.1668],
         [1.2611],
         [1.2125]]]), tensor([[[0.9924],
         [0.6173],
         [0.7373],
         [0.7168],
         [0.8335],
         [0.6937],
         [0.7781],
         [0.6942],
         [0.8155],
         [0.9057],
         [1.2075],
         [1.2914],
         [1.2661],
         [1.2761],
         [1.3702],
         [1.2854],
         [1.2459],
         [1.2902],
         [1.2528],
         [1.4128],
         [1.3326],
         [1.1793],
         [1.2480],
         [1.3251],
         [1.3396],
         [1.2936],
         [1.3273],
         [1.3298],
         [1.2488],
         [1.3414],
         [1.1576],
         [1.3409],
         [1.2262],
         [1.1738],
         [1.2447],
         [1.2755],
         [1.2298],
         [1.2371],
         [1.3093],
         [1.2824],
         [1.2545],
         [1.2782],
         [1.2779],
         [1.2605],
         [1.2905],
         [1.3088],
         [1.2622],
         [1.1889],
         [1.3104],
         [1.2604]]]), tensor([[[1.2008],
         [0.9651],
         [1.0532],
         [1.0746],
         [1.1106],
         [1.0001],
         [1.0416],
         [1.0289],
         [1.0634],
         [1.1073],
         [1.1696],
         [1.2710],
         [1.2270],
         [1.2533],
         [1.3381],
         [1.2204],
         [1.1941],
         [1.2550],
         [1.1384],
         [1.3560],
         [1.2639],
         [1.1397],
         [1.2323],
         [1.2798],
         [1.2779],
         [1.2446],
         [1.2584],
         [1.2945],
         [1.2255],
         [1.3072],
         [1.1283],
         [1.2578],
         [1.1830],
         [1.1192],
         [1.2179],
         [1.2350],
         [1.2123],
         [1.2121],
         [1.3214],
         [1.2791],
         [1.2262],
         [1.2775],
         [1.2705],
         [1.2618],
         [1.2535],
         [1.2882],
         [1.2268],
         [1.1871],
         [1.2990],
         [1.2367]]]))`

**Comment by Rtut654** [comment]
@amandeepbaberwal , have you succeed in running it on your own data?

**Comment by amandeepbaberwal** [comment]
@Rtut654 i was able to run it, but for some reason the scores don't change if we even use completely different audio for the sentence.

**Comment by Rtut654** [comment]
@amandeepbaberwal That sounds strange. I wonder if someone was able to infer the model with custom dataset.

**Comment by LyWangPX** [comment]
I met similar problems where the score did not change at all. I think the score is changing with the label instead of the data. Considering the bad quality of speechocean762, it might be a overfit or the dataset is just not practical.

I switched to the confidence of whisper model later which is more approachable.

**Comment by Rtut654** [comment]
@LyWangPX It's even more weird. Will try to find out why it is so. I tbh haven't seen a working repo for APED-like problems yet

**Comment by Rtut654** [comment]
@LyWangPX Hello! It is off the topic, but have you encounter similar issues with whsiper confidence? https://github.com/openai/whisper/discussions/1591

**Comment by gsabarinath02** [comment]
> I met similar problems where the score did not change at all. I think the score is changing with the label instead of the data. Considering the bad quality of speechocean762, it might be a overfit or the dataset is just not practical.
> 
> I switched to the confidence of whisper model later which is more approachable.

This sounds great, could u explain how to replace kaldi asr with whisper asr, and what changes need to be done ?
also, is it giving better results? or do u have any other asr model that is better than Kaldi asr ?
